src/ui/public/query_bar/components/query_bar.tsx

```diff
@@ -0,0 +1,502 @@
+    });
+  };
+
+  public incrementIndex = (currentIndex: number) => {
+    let nextIndex = currentIndex + 1;
+    if (currentIndex === null || nextIndex >= this.state.suggestions.length) {
+      nextIndex = 0;
+    }
+    this.setState({ index: nextIndex });
+  };
+
+  public decrementIndex = (currentIndex: number) => {
+    const previousIndex = currentIndex - 1;
+    if (previousIndex < 0) {
+      this.setState({ index: this.state.suggestions.length - 1 });
+    } else {
+      this.setState({ index: previousIndex });
+    }
+  };
+
+  public getSuggestions = async () => {
+    if (!this.inputRef) {
+      return;
+    }
+
+    const {
+      query: { query, language },
+    } = this.state;
+    const recentSearchSuggestions = this.getRecentSearchSuggestions(query);
+
+    const autocompleteProvider = getAutocompleteProvider(language);
+    if (!autocompleteProvider) {
+      return recentSearchSuggestions;
+    }
+
+    const indexPatterns = getFromLegacyIndexPattern(this.props.indexPatterns);
+    const getAutocompleteSuggestions = autocompleteProvider({ config, indexPatterns });
+
+    const { selectionStart, selectionEnd } = this.inputRef;
+    if (selectionStart === null || selectionEnd === null) {
+      return;
+    }
+
+    const suggestions: AutocompleteSuggestion[] = await getAutocompleteSuggestions({
+      query,
+      selectionStart,
+      selectionEnd,
+    });
+    return [...suggestions, ...recentSearchSuggestions];
+  };
+
+  public selectSuggestion = ({
+    type,
+    text,
+    start,
+    end,
+  }: {
+    type: AutocompleteSuggestionType;
+    text: string;
+    start: number;
+    end: number;
+  }) => {
+    if (!this.inputRef) {
+      return;
+    }
+
+    const query = this.state.query.query;
+    const { selectionStart, selectionEnd } = this.inputRef;
+    if (selectionStart === null || selectionEnd === null) {
+      return;
+    }
+
+    const value = query.substr(0, selectionStart) + query.substr(selectionEnd);
+
+    this.setState(
+      {
+        query: {
+          ...this.state.query,
+          query: value.substr(0, start) + text + value.substr(end),
+        },
+        index: null,
+      },
+      () => {
+        if (!this.inputRef) {
+          return;
+        }
+
+        this.inputRef.setSelectionRange(start + text.length, start + text.length);
+
+        if (type === recentSearchType) {
+          this.onSubmit();
+        } else {
+          this.updateSuggestions();
+        }
+      }
+    );
+  };
+
+  public getRecentSearchSuggestions = (query: string) => {
+    if (!this.persistedLog) {
+      return [];
+    }
+    const recentSearches = this.persistedLog.get();
+    const matchingRecentSearches = recentSearches.filter(recentQuery => {
+      const recentQueryString = typeof recentQuery === 'object' ? toUser(recentQuery) : recentQuery;
+      return recentQueryString.includes(query);
+    });
+    return matchingRecentSearches.map(recentSearch => {
+      const text = recentSearch;
```

Comment:
@Bargs, you should also call toUser() on recentSearch if it is an object.  Otherwise, react doesn't like it when the object is used in the Suggestion component.